src/NServiceBus.Core/Recoverability/SecondLevelRetries/SecondLevelRetriesBehavior.cs

```diff
@@ -15,16 +15,40 @@ namespace NServiceBus
 
     class SecondLevelRetriesBehavior : Behavior<TransportReceiveContext>
     {
-        public SecondLevelRetriesBehavior(IPipelineBase<RoutingContext> dispatchPipeline, SecondLevelRetryPolicy retryPolicy, BusNotifications notifications, string localAddress)
+        public SecondLevelRetriesBehavior(IPipelineBase<RoutingContext> dispatchPipeline, SecondLevelRetryPolicy retryPolicy, BusNotifications notifications, string localAddress, SlrStatusStorage slrStatusStorage)
         {
             this.dispatchPipeline = dispatchPipeline;
             this.retryPolicy = retryPolicy;
             this.notifications = notifications;
             this.localAddress = localAddress;
+            this.slrStatusStorage = slrStatusStorage;
         }
 
         public override async Task Invoke(TransportReceiveContext context, Func<Task> next)
         {
+            Exception exception;
+            var uniqueMessageId = PipelineInfo.Name + context.Message.MessageId;
+
+            if (slrStatusStorage.TryGetException(uniqueMessageId, out exception))
+            {
+                slrStatusStorage.ClearException(uniqueMessageId);
+
+                TimeSpan delay;
+                int currentRetry;
+
+                if (ShouldPerformSlr(context.Message, exception, out delay, out currentRetry))
+                {
+                    await QueueForDelayedDelivery(context, currentRetry, delay, exception);
+
+                    return;
```

Comment:
I thought that was a good catch but after checking it turns out that it's not allowed for async methods which return Task and not Task<T>. So probably needs to stay the way it is now.